### PR TITLE
[camera_windows] Support image streams on Windows platform

### PR DESCRIPTION
**Pull request by jlundOverlay**
Adds support to the windows camera implementation to allow streaming of frames.

Fixes https://github.com/flutter/flutter/issues/97542

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

### CODE REVIEW

**Comment by cbracken** [comment]
Taking a look, but before I do, looks like some small Dart analysis errors:
```
   info - lib/camera_windows.dart:9:8 - Use relative imports for files in the 'lib' directory. Try converting the URI to a relative URI. - prefer_relative_imports
   info - lib/type_conversion.dart:19:15 - Missing type annotation. Try adding a type annotation. - always_specify_types
   info - lib/type_conversion.dart:22:24 - Unnecessary use of parentheses. Try removing the parentheses. - unnecessary_parenthesis
```

File: packages/camera/camera_windows/windows/capture_controller.h
Comment (cbracken) [inline packages/camera/camera_windows/windows/capture_controller.h:16]: nit: Order includes alphabetically within groups (move this just above the `flutter/texture_register.h` include.

File: packages/camera/camera_windows/windows/capture_controller.h
Comment (cbracken) [inline packages/camera/camera_windows/windows/capture_controller.h:118]: Add short doc comments to these two methods, similar to the ones above.

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (cbracken) [inline packages/camera/camera_windows/windows/camera_plugin.cpp:609]: nit: here and elsewhere, comments should follow the Flutter style guide.

Ref: https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#documentation-dartdocs-javadocs-etc

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (cbracken) [inline packages/camera/camera_windows/windows/camera_plugin.cpp:294]:
```suggestion
  } else if (method_name.compare(kStopImageStreamingMethod) == 0) {
```

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (cbracken) [inline packages/camera/camera_windows/windows/camera_plugin.cpp:300]:
```suggestion
  } else if (method_name.compare(kPausePreview) == 0) {
```

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (cbracken) [reply packages/camera/camera_windows/windows/camera_plugin.cpp:609]: Here and in the stop method: Given that `camera_id` is a pointer, please make the type explicit to help readers rather than using `auto`.

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (cbracken) [inline packages/camera/camera_windows/windows/camera_plugin.cpp:615]: Similar comment here (and at the equivalent spot in the stop method); please make the type explicit.

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (cbracken) [inline packages/camera/camera_windows/windows/camera_plugin.cpp:636]: nit: Delete blank lines here.

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (cbracken) [inline packages/camera/camera_windows/windows/camera_plugin.cpp:638]: nit: add a blank line here.

File: packages/camera/camera_windows/windows/capture_controller.cpp
Comment (cbracken) [inline packages/camera/camera_windows/windows/capture_controller.cpp:14]: nit: Move this up a line to group it with the other C++ includes.

File: packages/camera/camera_windows/windows/capture_controller.cpp
Comment (cbracken) [inline packages/camera/camera_windows/windows/capture_controller.cpp:24]: nit: C includes should go up with the other C includes (`comdef.h`, `wincodec.h`, `wrl/client.h`).

File: packages/camera/camera_windows/windows/capture_controller.cpp
Comment (cbracken) [inline packages/camera/camera_windows/windows/capture_controller.cpp:897]:
```suggestion
      // Convert the buffer data to a std::vector<uint8_t>.
```

File: packages/camera/camera_windows/windows/capture_controller.cpp
Comment (cbracken) [inline packages/camera/camera_windows/windows/capture_controller.cpp:900]:
```suggestion
      // Ensure preview_frame_height_ and preview_frame_width_ are of supported types.
```

File: packages/camera/camera_windows/windows/capture_controller.cpp
Comment (cbracken) [inline packages/camera/camera_windows/windows/capture_controller.cpp:904]:
```suggestion
      // Create a map to hold the buffer data and data length.
```

File: packages/camera/camera_windows/windows/capture_controller.cpp
Comment (cbracken) [inline packages/camera/camera_windows/windows/capture_controller.cpp:911]:
```suggestion
      // Wrap the map in a flutter::EncodableValue.
```

File: packages/camera/camera_windows/windows/capture_controller.cpp
Comment (cbracken) [inline packages/camera/camera_windows/windows/capture_controller.cpp:914]:
```suggestion
      // Send the encoded value through the image_stream_sink_.
```

File: packages/camera/camera_windows/windows/capture_controller.cpp
Comment (cbracken) [inline packages/camera/camera_windows/windows/capture_controller.cpp:916]: nit: formatting.

**Review by cbracken — CHANGES_REQUESTED** [review]
Thanks for your PR! A few first-round comments.

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (jlundOverlay) [reply packages/camera/camera_windows/windows/camera_plugin.cpp:609]: will fix the comments, happy to explicitly type the camera_id however I had left it as auto for consistency with the other pre existing functions.

Comment (cbracken) [reply packages/camera/camera_windows/windows/camera_plugin.cpp:609]: Thanks -- please do. The original camera plugin was written in a relatively short period and not particularly in a style consistent with the rest of Flutter's code/packages.

Generally, when it comes to `auto` we should be following the rules in the Google C++ style guide unless explicitly overridden by our own style guide. When it comes to type, the advice is:

"Use type deduction only if it makes the code clearer to readers who aren't familiar with the project, or if it makes the code safer. Do not use it merely to avoid the inconvenience of writing an explicit type."

The existing code definitely overuses `auto` -- apologies for the confusion.

Ref: https://google.github.io/styleguide/cppguide.html#Type_deduction

Comment (jlundOverlay) [reply packages/camera/camera_windows/windows/camera_plugin.cpp]: No worries, I have made the changes to the functions(happy to replicate this to the other similar functions if you are happy with it?). also should the arguments on 289 & 295 be explict?

Comment (cbracken) [reply packages/camera/camera_windows/windows/camera_plugin.cpp]: Thanks for checking! For `std::get_if<T>` the type being returned is clear from the template type so adding it to the left hand side doesn't add any more clarity, so I think those two lines are a good example of where to use `auto` and avoid repeating the type.

**Comment by jlundOverlay** [comment]
Looks like the camera windows implementation was just switched to pigeon, unfortunate timing but I'll migrate the changes across

**Comment by cbracken** [comment]
Ugh yeah was just thinking about that as I reviewed the other one. On the plus side, the code should be significantly simpler, but frustrating to deal with the merge.

**Comment by jlundOverlay** [comment]
I think i have worked out how to replace the method channels, just unsure if i still need the frame event channel. https://github.com/flutter/flutter/issues/66711 seems to indicate that pigeon doesn't use an event channel under the hood so I'm assuming its less performant for this use case?

**Comment by sneurlax** [comment]
I tested this as working :) using @jlundOverlay's fork until this is merged, thanks!

**Review by cbracken — APPROVED** [review]
lgtm for the change, modulo bot failures:
* please add a changelog entry.
* please fix formatting errors reported in the test failure.

LGTM stamp from a Japanese personal seal

**Comment by jlundOverlay** [comment]
should be good to go assuming I got the right version numbers😅

**Review by cbracken — APPROVED** [review]
Looks good to me!

**Comment by cbracken** [comment]
@bparrishMines do you want to be the second reviewer? If so, needs the approval button.

File: packages/camera/camera/CHANGELOG.md
Comment (loic-sharma) [inline packages/camera/camera/CHANGELOG.md:3]: This change updates both `camera` and `camera_windows` in a single pull request. Don't we need to split this into two separate pull requests?

See: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changing-federated-plugins

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (loic-sharma) [inline packages/camera/camera_windows/windows/camera_plugin.cpp:410]: Should this be `PendingResultTYpe::kStopStream`?

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (loic-sharma) [inline packages/camera/camera_windows/windows/camera_plugin.cpp:403]: Should this be `PendingREsultType::kStopStream`?

```suggestion
  }
  if (camera->HasPendingResultByType(PendingResultType::kStopStream)) {
    return result(
        FlutterError("camera_error", "Pending stop stream request exists"));
  }
```

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (loic-sharma) [inline packages/camera/camera_windows/windows/camera_plugin.cpp:405]: Could you expand why this check is necessary? My understanding is that the event sink is moved to `CameraController::StartImageStream` - shouldn't this `event_sink` be `nullptr` at this point?

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (jlundOverlay) [reply packages/camera/camera_windows/windows/camera_plugin.cpp:410]: yep definitely, good pickup

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (jlundOverlay) [reply packages/camera/camera_windows/windows/camera_plugin.cpp]: I think you're right, i have removed the check

**Comment by jlundOverlay** [comment]
I have made a new PR for the version bump https://github.com/flutter/packages/pull/7220

File: packages/camera/camera/CHANGELOG.md
Comment (jlundOverlay) [reply packages/camera/camera/CHANGELOG.md]: pretty confusing steps, but i have split the PR

**Comment by loic-sharma** [comment]
@jlundOverlay Ideally we should have tests for this change. What do you think of updating `packages/camera/camera_windows/windows/test/camera_plugin_test.cpp` to check that `CameraPlugin::StartImageStream` and `CameraPlugin::StopImageStream` interact correctly with the `CaptureController`?

File: packages/camera/camera/CHANGELOG.md
Comment (loic-sharma) [reply packages/camera/camera/CHANGELOG.md]: Thanks! Agreed it's a little confusing, that's the small drawback for auto publishing packages :)

**Comment by jlundOverlay** [comment]
Took me a bit to work out how to mock the EventSink but i think i got there in the end. let me know what you think.

**Review by loic-sharma — APPROVED** [review]
Looks great! Excellent work!

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (stuartmorgan) [inline packages/camera/camera_windows/windows/camera_plugin.cpp:390]: Where are the pending results added in this PR resolved? I don't see anything in the diff that retrieves them.

File: packages/camera/camera_windows/pigeons/messages.dart
Comment (stuartmorgan) [inline packages/camera/camera_windows/pigeons/messages.dart:74]: Why are these `@async`? It looks like the logic is synchronous.

File: packages/camera/camera_windows/windows/camera_plugin.cpp
Comment (jlundOverlay) [reply packages/camera/camera_windows/windows/camera_plugin.cpp:390]: good question, when i wrote it figured id deal with that later and forgot. What exactly are these checks used for and are they necessary if the function is synchronous?

Comment (stuartmorgan) [reply packages/camera/camera_windows/windows/camera_plugin.cpp:390]: There will be no result object if the API is made synchronous.

**Comment by yoongyo** [comment]
Please keep making progress.

**Pull request merged**
